File: content-solutions/graph-representation/graph-grid-paths.java
class Solution {
    public int[][] solution(String[] grid) {
        int rows = grid.length, cols = grid[0].length();
        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        int[][] answer = new int[rows][cols];
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r].charAt(c) == '#') {
                    answer[r][c] = -1;
                    continue;
                }
                int count = 0;
                for (int[] d : dirs) {
                    int nr = r + d[0], nc = c + d[1];
                    if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr].charAt(nc) == '.') count++;
                }
                answer[r][c] = count;
            }
        }
        return answer;
    }
}

File: content-solutions/graph-representation/graph-grid-paths.js
function solution(grid) {
  const rows = grid.length;
  const cols = grid[0].length;
  const dirs = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];
  return grid.map((line, r) =>
    [...line].map((cell, c) => {
      if (cell === "#") return -1;
      let count = 0;
      for (const [dr, dc] of dirs) {
        const nr = r + dr;
        const nc = c + dc;
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] === ".") count += 1;
      }
      return count;
    }),
  );
}
